### PR TITLE
Allow preprocessors to specify extra C sources

### LINKED ISSUE #238: Cabal should support hsc2hs's stub .c feature
(Imported from [Trac #245](http://hackage.haskell.org/trac/hackage/ticket/245), reported by guest on 2008-02-21)

The Pugs project recently uncovered a bit of a problem with GHC 6.8.x.

To quote:

- {-# INCLUDE -} seems to require either a very relative or an absolute path. That is, you have to make the path listed inside an INCLUDE ('dist/build/Pugs/Embed/Parrot_hsc.h') in the file dist/build/Pugs/Embed/Parrot.hs absolute or completely relative - that is, the first line needs to look like either '{-# INCLUDE "Parrot_hsc.h" #-}' or  '{-# INCLUDE

<blockquote>
"/home/gwern/bin/pugs/dist/build/Pugs/Embed/Parrot_hsc.h" #-}'.

</blockquote>
The particular problem is that  'dist/build/Pugs/Embed/Parrot_hsc.c:#include "Parrot_hsc.h"' is being translated to the (broken) 'dist/build/Pugs/Embed/Parrot.hs:{-# INCLUDE "dist/build/Pugs/Embed/Parrot_hsc.h" #-}'.

The relevant files and notes can be found in the Pugs SVN repo, in src/Pugs/Embed/, and the files of interest are Haskell.hs  Parrot.hsc  Parrot_hsc.c  Parrot_hsc.h  Perl5.hs  Pugs.hs

--
gwern

### PR DESCRIPTION
**Pull request by ian-ross**
Alter return type from preprocessing to allow preprocessors like hsc2hs and C2HS to inform Cabal of extra C sources that they create that need to be compiled and linked.  I believe this fixes #238 as well as providing equivalent support for a related feature in new versions of C2HS where the preprocessor generates some C wrapper functions to help with argument marshalling.

### CODE REVIEW

**Comment by 23Skidoo** [comment]
This breaks the build on Travis.

Please add a changelog note. Maybe also a test and some documentation?

**Comment by ian-ross** [comment]
@23Skidoo Will do.  I hadn't seen the warnings that are breaking the Travis build.  And where should documentation go?  Just inline comments, or is there a better place?  (Sorry if that's a dumb question: this is the first time I've contributed to Cabal...)

**Comment by 23Skidoo** [comment]
You probably compiled without `-Wall`. Documentation for user-facing features [lives under `Cabal/doc`](https://github.com/haskell/cabal/tree/master/Cabal/doc).

**Comment by ian-ross** [comment]
@23Skidoo I think this is now better -- there's a test (that helped fix a problem...), a changelog comment, some documentation and the build on Travis is fine.  What's your preferred approach for merging PRs with extra commits?  Should I squash the commits into one and force push, close the PR and open another one with a squashed commit, or just leave it as is?

**Comment by 23Skidoo** [comment]
> Should I squash the commits into one and force push, close the PR and open another one with a squashed commit, or just leave it as is?

If you want to squash the commits, you can just do `push -f` after `rebase -i`. The PR will be updated. This sometimes makes code review comments attached to old diffs (in this case there aren't any) to vanish, though.

**Comment by ian-ross** [comment]
OK, rebased and pushed.

File: cabal-install/Distribution/Client/SetupWrapper.hs [hunk 1]
```diff
@@ -333,7 +333,9 @@ externalSetupMethod verbosity options pkg bt mkargs = do
     srcNewer <- src `moreRecentFile` setupHs
     when srcNewer $ if useHs
                     then copyFileVerbose verbosity src setupHs
-                    else runSimplePreProcessor ppUnlit src setupHs verbosity
+                    else do
```
Comment (23Skidoo) [inline cabal-install/Distribution/Client/SetupWrapper.hs]: Isn't this do-block equivalent to `void $ runSimplePreProcessor ...`?

File: Cabal/Distribution/Simple/PreProcess.hs [hunk 2]
```diff
@@ -126,20 +126,29 @@ data PreProcessor = PreProcessor {
...
                       -> (FilePath, FilePath)
-                      -> (FilePath, FilePath) -> Verbosity -> IO ()
+                      -> (FilePath, FilePath) -> Verbosity -> IO [FilePath]
 mkSimplePreProcessor simplePP
   (inBaseDir, inRelativeFile)
+  (outBaseDir, outRelativeFile) verbosity = simplePP inFile outFile verbosity >> return []
+  where inFile  = normalise (inBaseDir  </> inRelativeFile)
+        outFile = normalise (outBaseDir </> outRelativeFile)
+
+mkLessSimplePreProcessor :: (FilePath -> FilePath -> Verbosity -> IO [FilePath])
```
Comment (23Skidoo) [inline Cabal/Distribution/Simple/PreProcess.hs]: Not a very good name, IMO. Maybe add a comment explaining why it is "less simple".

**Comment by 23Skidoo** [comment]
LGTM with minor comments.

This changes the `PreProcessor` type, and by extension `PPSuffixHandler`, which is part of `UserHooks`. @dcoutts, are we OK with this?

File: cabal-install/Distribution/Client/SetupWrapper.hs (hunk 1 above)
Comment (ian-ross) [inline cabal-install/Distribution/Client/SetupWrapper.hs]: Oops.  Yes.  I'll fix it.

File: Cabal/Distribution/Simple/PreProcess.hs (hunk 2 above)
Comment (ian-ross) [inline Cabal/Distribution/Simple/PreProcess.hs]: It's a *terrible* name.  It was supposed to be a placeholder while I thought of a better name...  I'll change it and add a comment.

**Comment by 23Skidoo** [comment]
So the main problem here is that this change breaks Setup scripts that define custom preprocessors. See [here](http://stackoverflow.com/questions/7370029/any-example-of-a-custom-preprocessor-in-haskell) for an example of such setup script.

**Comment by ian-ross** [comment]
@23Skidoo OK, let me think of a way to do this in a way that's backwards compatible with the existing `PreProcessor` types.  I didn't really take on board that that stuff was user-visible.

**Comment by 23Skidoo** [comment]
This may be quite hard given that the `UserHooks` API is not really [designed for extensibility](http://www.yesodweb.com/blog/2015/03/designing-apis-for-extensibility). If we ever do a Cabal 2.0 release, we should take the opportunity to hide the internal structure of all user-visible records.

**Comment by ian-ross** [comment]
Yeah, I was just looking at that.  It's going to be hard to do this, because the `PreProcessor` type is exported as `PreProcessor(..)` -- I changed `mkSimplePreProcessor` in a way that would work with that example you linked to, but if people can construct arbitrary `PreProcessor` instances, I can't see how to deal with a change in the type of `runPreProcessor` in a backwards compatible way.

**Comment by 23Skidoo** [comment]
There should at least be a way to write setup scripts that are compatible with both old and new versions of Cabal. It'd be also nice to check whether this change breaks any currently existing package on Hackage.

**Comment by ian-ross** [comment]
That makes sense.  Let me collect some data: I'll look for all the packages on Hackage that use custom build scripts and see if any of them are broken by this change.  I'll also collect all the examples of custom preprocessors I can find.  That won't help with peoples' private code that's not on Hackage, but at least it's a start and should give some idea of how this is used in practice.  Results in a couple of days...

**Comment by 23Skidoo** [comment]
Cool, thank you for being so thorough.

**Comment by ian-ross** [comment]
OK, this would break all the Gtk packages.  Don't see a way to get round it.

**Comment by ian-ross** [comment]
@23Skidoo OK, I've now done this in a different way.  It's kind of ugly, but it doesn't change the user-visible preprocessor API at all, so it ought not to perturb any of the existing packages that use preprocessors in custom builds.  Given the constraints on changes to the public Cabal API, this seems like a better way to do things for now.

**Comment by ian-ross** [comment]
@23Skidoo Thanks for the detailed review!  I've added Haddock comments, added full module names in the error messages, and have changed the way that the extra C sources are searched for, just looking for files matching the relevant pattern under the build directory (which is where these things end up, not in the autogen directory).

**Comment by 23Skidoo** [comment]
Thanks, everything looks good to me now. Merging.

**Comment by ian-ross** [comment]
Great.  That was kind of epic.  Thanks for being patient.

**Comment by 23Skidoo** [comment]
Thanks to you for your contribution!

**Pull request merged**